src/reports_updater.rs

```diff
@@ -96,45 +104,242 @@ impl TcnMatcherRayon {
     }
 }
 
-#[derive(Debug, Eq, PartialEq, Clone)]
+#[derive(Debug, PartialEq, Clone)]
 pub struct ObservedTcn {
     tcn: TemporaryContactNumber,
-    time: UnixTime,
+    contact_start: UnixTime,
+    contact_end: UnixTime,
+    min_distance: f32,
+    avg_distance: f32,
+    total_count: usize // Needed to calculate correctly average of averages (= average of single values)
 }
 
 pub trait ObservedTcnProcessor {
-    fn save(&self, tcn_str: &str) -> Result<(), ServicesError>;
+    fn save(&self, tcn_str: &str, distance: f32) -> Result<(), ServicesError>;
 }
 
-pub struct ObservedTcnProcessorImpl<T>
+#[derive(Copy, Clone)]
+struct Element {}
+
+pub struct TcnBatchesManager<T>
 where
     T: TcnDao,
 {
-    pub tcn_dao: Arc<T>,
+    tcn_dao: Arc<T>,
+    tcns_batch: Mutex<HashMap<[u8; 16], ObservedTcn>>,
+    exposure_grouper: ExposureGrouper,
+}
+
+impl<T> TcnBatchesManager<T>
+where
+    T: 'static + TcnDao,
+{
+    pub fn new(tcn_dao: Arc<T>, exposure_grouper: ExposureGrouper) -> TcnBatchesManager<T> {
+        TcnBatchesManager {
+            tcn_dao,
+            tcns_batch: Mutex::new(HashMap::new()),
+            exposure_grouper,
+        }
+    }
+
+    pub fn flush(&self) -> Result<(), ServicesError> {
+        let res = self.tcns_batch.lock();
```

Comment:
I don't like this. `flush` is executed in a background thread, but the lock will block anyway the threads pushing the TCNs. The later seem to belong to BLE, at least on iOS.

Reply:
Done: https://github.com/Co-Epi/app-backend-rust/pull/123/commits/f2304da8059d0f691ad3ebf6f67095042af264f8 / https://github.com/Co-Epi/app-backend-rust/pull/123/commits/82f19dd4939d380884f69879b03d28c2330dc753